proc_smaps.go

```diff
@@ -0,0 +1,246 @@
+	MMUPageSize uint64
+	// Kernel flags associated with the virtual memory area
+	VMFlags string
+}
+
+// SizeSum returns the sum of Pss from all mappings
+func (s ProcSMaps) SizeSum() (sum uint64) {
+	for _, x := range s {
+		sum += x.Size
+	}
+
+	return
+}
+
+// RssSum returns the sum of Pss from all mappings
+func (s ProcSMaps) RssSum() (sum uint64) {
+	for _, x := range s {
+		sum += x.Rss
+	}
+
+	return
+}
+
+// PssSum returns the sum of Pss from all mappings
+func (s ProcSMaps) PssSum() (sum uint64) {
+	for _, x := range s {
+		sum += x.Pss
+	}
+
+	return
+}
+
+// SharedCleanSum returns the sum of Pss from all mappings
+func (s ProcSMaps) SharedCleanSum() (sum uint64) {
+	for _, x := range s {
+		sum += x.SharedClean
+	}
+
+	return
+}
+
+// SharedDirtySum returns the sum of Pss from all mappings
+func (s ProcSMaps) SharedDirtySum() (sum uint64) {
+	for _, x := range s {
+		sum += x.SharedDirty
+	}
+
+	return
+}
+
+// PrivateCleanSum returns the sum of Pss from all mappings
+func (s ProcSMaps) PrivateCleanSum() (sum uint64) {
+	for _, x := range s {
+		sum += x.PrivateClean
+	}
+
+	return
+}
+
+// PrivateDirtySum returns the sum of Pss from all mappings
+func (s ProcSMaps) PrivateDirtySum() (sum uint64) {
+	for _, x := range s {
+		sum += x.PrivateDirty
+	}
+
+	return
+}
+
+// ReferencedSum returns the sum of Pss from all mappings
+func (s ProcSMaps) ReferencedSum() (sum uint64) {
+	for _, x := range s {
+		sum += x.Referenced
+	}
+
+	return
+}
+
+// AnonymousSum returns the sum of Pss from all mappings
+func (s ProcSMaps) AnonymousSum() (sum uint64) {
+	for _, x := range s {
+		sum += x.Anonymous
+	}
+
+	return
+}
+
+// SwapSum returns the sum of Pss from all mappings
+func (s ProcSMaps) SwapSum() (sum uint64) {
+	for _, x := range s {
+		sum += x.Swap
+	}
+
+	return
+}
+
+// SwapPssSum returns the sum of Pss from all mappings
+func (s ProcSMaps) SwapPssSum() (sum uint64) {
+	for _, x := range s {
+		sum += x.SwapPss
+	}
+
+	return
+}
+
+// ProcSMaps reads from /proc/[pid]/maps to get the memory-mappings of the
+// process.
+func (p Proc) ProcSMaps() (ProcSMaps, error) {
+	file, err := os.Open(p.path("smaps"))
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	var currentMap *ProcSMap
+
+	smaps := []*ProcSMap{}
+	scan := bufio.NewScanner(file)
+
+	for scan.Scan() {
+		line := scan.Text()
+		// first line of a mapping start with an hexadecimal address in lower-case
+		// All other line start with a capitalized words.
+		if (line[0] >= '0' && line[0] <= '9') || (line[0] >= 'a' && line[0] <= 'f') {
+			if currentMap != nil {
+				smaps = append(smaps, currentMap)
+			}
+
+			m, err := parseProcMap(line)
+			if err != nil {
+				return nil, err
+			}
+
+			currentMap = &ProcSMap{
+				ProcMap: m,
+			}
+
+			continue
+		}
+
+		kv := strings.SplitN(line, ":", 2)
+		if len(kv) != 2 {
+			continue
+		}
+
+		k := kv[0]
+		v := strings.TrimSpace(kv[1])
+		v = strings.TrimRight(v, " kB")
+
+		vKBytes, _ := strconv.ParseUint(v, 10, 64)
```

Comment:
It might be better to check this error and fail if it's non-nil unless there is reason to ignore the parse failures.